haskell/set.bzl

```diff
@@ -0,0 +1,111 @@
+"""Immutable sets that support efficient merging, traversal, and membership
+check.
+"""
+
+def _new():
+  """Create an empty set.
+  """
+  return struct(_set_items = dict())
+
+def _is_member(s, e):
+  """Return true if `e` is in the set `s`.
+
+  Args:
+    s: The set to inspect.
+    e: The element to search for.
+
+  Result:
+    Bool, true if `e` is in `s`, false otherwise.
+  """
+  e in s._set_items.get(e)
+
+def _insert(s, e):
+  """Insert an element into the set.
+
+  Args:
+    s: Set to insert new element into.
+    e: The element to insert.
+
+  Result:
+    A copy of set `s` with `s` element added.
+  """
+  r = dict(s._set_items)
+  r[e] = None
+  return struct(_set_items = r)
+
+def _union(s0, s1):
+  """Return union of two sets.
+
+  Args:
+    s0: One set.
+    s1: Another set.
+
+  Result:
+    Set, union of the two sets.
+  """
+  r = dict(s0._set_items)
+  r.update(s1._set_items)
+  return struct(_set_items = r)
+
+def _map(s, f):
```

Comment:
Considering lack of lambda &c., I think this is pretty much useless. I'm not even sure bazel lets you pass a function in to begin with actually.

If we're keeping it, use dict comprehension instead, IIRC it can be a lot faster.

```
return struct(_set_items = { f(x): None for x in s_set_items.keys() })
```

Reply:
Ah I see it used below actually, disregard first part then.